text/0000-inline-asm.md

```diff
@@ -0,0 +1,1002 @@
+{
+    let c;
+    {
+        #[repr(C)]
+        struct foo_inline_asm_args {
+            a: i32,
+            b: i32,
+            c: i32,
+        }
+        extern "C" {
+            fn foo_inline_asm(args: *mut foo_inline_asm_args);
+        }
+        let mut args = foo_inline_asm_args {
+            a: a,
+            b: b,
+            c: mem::uninitialized(),
+        };
+        foo_inline_asm(&mut args);
+        a = args.a;
+        c = args.c;
+    }
+    (a, c)
+}
+```
+
+## Rules for inline assembly
+[rules]: #rules
+
+- Any registers not specified as inputs will contain an undefined value on entry to the asm block.
+  - An "undefined value" in this context means that the register can have any one of the possible values allowed by the architecture. Notably it is not the same as an LLVM `undef` which can have a different value every time you read it (since such a concept does not exist in assembly code).
+- Any registers not specified as outputs must have the same value upon exiting the asm block as they had on entry, otherwise behavior is undefined.
+  - This only applies to registers which can be specified as an input or output. Other registers follow target-specific rules and are outside the scope of this RFC.
+  - Note that a `lateout` may be allocated to the same register as an `in`, in which case this rule does not apply. Code should not rely on this however since it depends on the results of register allocation.
+- Behavior is undefined if execution unwinds out of an asm block.
+- Any memory reads/writes performed by the asm code follow the same rules as `volatile_read` and `volatile_write`.
+  - Refer to the unsafe code guidelines for the exact rules.
+  - If the `readonly` option is set, then only memory reads (with the same rules as `volatile_read`) are allowed.
+  - If the `nomem` option is set then no reads or write to memory are allowed.
+  - These rules do not apply to memory which is private to the asm code, such as stack space allocated within the asm block.
+- The compiler cannot assume that the instructions in the asm are the ones that will actually end up executed.
+  - This effectively means that the compiler must treat the `asm!` as a black box and only take the interface specification into account, not the instructions themselves.
+  - Runtime code patching is allowed, via target-specific mechanisms (outside the scope of this RFC).
+- Unless the `nostack` option is set, asm code is allowed to use stack space below the stack pointer.
+  - On entry to the asm block the stack pointer is guaranteed to be suitably aligned (according to the target ABI) for a function call.
+  - You are responsible for making sure you don't overflow the stack (e.g. use stack probing to ensure you hit a guard page).
+  - You should adjust the stack pointer when allocating stack memory as required by the target ABI.
+  - The stack pointer must be restored to its original value before leaving the asm block.
```

Comment:
> The stack pointer must be restored to its original value before leaving the asm block.

It just occured to me, does this effectively disallows scheduler/context switching from an asm block? It looks like this disallows schemes where we save esp (along with the rest of the machine context) to switch to another "thread"/execution context. The issue is that when process switching, we save the current esp, and restore the esp of another task, resuming that task. When leaving the asm block, we'll have a different esp from the one we entered (although it's still valid from the POV of that function). See [this](https://github.com/sunriseos/SunriseOS/blob/master/kernel/src/i386/process_switch.rs#L155) as an example of a process switch written with the old asm.

Reply:
I believe this has previously been stated as acceptable because you're returning from a different `asm!` block.

Reply:
I believe LLVM can move code around the asm block, so the mentioned example is likely incorrect.

Reply:
Have you seen the notes I added in d0a743a99bf14d91a3d29e946aa7fa9af9af04ac about context-switching?

Reply:
Oh damn really sorry, I must have been reading an old version of the RFC before leaving this comment.

However, if what @bjorn3 says is true (FWIW I haven't seen it happen in my code, but as all optimization-related things YMMV), then it looks like there are additional requirements around code that changes ESP? I was under the impression that making it `volatile` would prevent llvm from reordering the code happening before the asm, and the code happening after the asm depends on the output from the asm block anyways, so that shouldn't even be possible to reorder.

Reply:
I think you're just missing a `"memory"` clobber in your code. What is more concerning is your use of `current_esp`: you are assuming that the compiler will only update the stack pointer in the prologue/epilogue of `process_switch`, but this isn't actually guaranteed. I am aware of some compiler optimizations which defer allocating a stack frame until later, e.g. if it is only needed for the slow path of a function. You really should be writing to `thread_current_lock_phwcontext.esp` from within the main `asm!` block. Anyways this conversation is getting off-topic for this RFC, so let's stop here.